### PR TITLE
Denote if a printer is generic.

### LINKED ISSUE #38779: Update kubectl.GetPrinter to not return both nil for the printer and nil for the error on `wide` format
**Problem**

Right now, printer helpers such as [cmdutil.PrinterForCommand](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/printing.go#L128) call [kubectl.GetPrinter](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L74) with the output format specified by a `--output` flag, in order to retrieve the right printer for that format.

When `kubect.GetPrinter` is called with a `wide` format, [nil is returned for both the printer and the error](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L144). This has inevitably led to panics throughout kubectl commands whenever code that calls `kubectl.GetPrinter`  (either directly or indirectly through cmdutil.PrinterForCommand`) attempts to use the retrieved printer if no errors are returned. For example:

```go
// $ kubectl ... -o wide

outputFormat := cmdutil.GetFlagString(cmd, "output")
printer, generic, err := kubectl.GetPrinter(outputFormat, ...)
if err != nil {
    // handle error
    return
}

// any code here assumes no error AND that the printer exists, however since both the printer and the 
// error were nil for the "wide" outputFormat, the line below will cause a panic
printer.PrintObj(info.Object, out)
```

Commands like `kubectl create` and `kubectl apply` get around this by [only supporting the `name` output format](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create.go#L156), however when expanding these commands [to eventually use all supported output formats by resource printers](https://github.com/kubernetes/kubernetes/pull/38112) (json, yaml, custom-columns, etc...), this becomes an issue.

**Proposal**

1. A solution to this could either be to actually return an error in [kubect.GetPrinter](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L144) if the returned printer will be `nil`
 - which would then allow existing helpers such as [cmdutil.PrinterForCommand](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/printing.go#L108) to handle this and return a `NewHumanReadablePrinter` with the right `PrintOptions` for example, 

2. or to simply return a `NewHumanReadablePrinter` in `kubectl.GetPrinter` any time the output format specified == `wide`.

cc @fabianofranz @pwittrock @AdoHe @deads2k

### PR DESCRIPTION
**Pull request by waseem**
This fixes #38779.

This allows us to avoid case in which printers.GetStandardPrinter
returns nil for both printer and err removing any potential panics that
may arise throughout kubectl commands.

Please see #38779 and #38112 for complete context.

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 1]
```diff
@@ -74,25 +92,12 @@ func (f *ring2Factory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTM
...
-			ShowAll:            GetFlagBool(cmd, "show-all"),
-			ShowLabels:         GetFlagBool(cmd, "show-labels"),
-			AbsoluteTimestamps: isWatch(cmd),
-			ColumnLabels:       columnLabel,
-		})
-		if err != nil {
-			return nil, err
+		if humanReadablePrinter, ok := printer.(*printers.HumanReadablePrinter); ok {
+			printersinternal.AddHandlers(humanReadablePrinter)
+			printer = printers.ResourcePrinter(humanReadablePrinter)
```
Comment (droot) [inline pkg/kubectl/cmd/util/factory_builder.go]: Why is this line needed ? (may be a comment explaining the above two lines will help the reader here).

**Review by droot — COMMENTED** [review]
LGTM.

File: pkg/kubectl/cmd/util/factory_builder.go (hunk 1 above)
Comment (waseem) [inline pkg/kubectl/cmd/util/factory_builder.go]: @droot I added a comment explaining the need of these lines. While writing the comment I realized that I was doing an unnecessary conversion which I removed. PTAL.

Can you please queue the test runs again and also remove the `release-note-label-needed` label?

Thanks!

File: pkg/kubectl/cmd/get.go [hunk 2]
```diff
@@ -513,3 +513,7 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 	cmdutil.PrintFilterCount(errOut, len(objs), filteredResourceCount, len(allErrs), filterOpts, options.IgnoreNotFound)
 	return utilerrors.NewAggregate(allErrs)
 }
+
+func shouldGetNewPrinterForMapping(printer printers.ResourcePrinter, lastMapping, mapping *meta.RESTMapping) bool {
+	return printer == nil || lastMapping == nil || mapping == nil || mapping.Resource != lastMapping.Resource
+}
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/get.go]: +1 to this change

**Review by alexandercampbell — APPROVED** [review]

**Comment by alexandercampbell** [comment]
LGTM

**Comment by mengqiy** [comment]
@k8s-bot ok to test
/release-note-none

File: pkg/kubectl/cmd/util/factory_builder.go (hunk 1 above)
Comment (droot) [inline pkg/kubectl/cmd/util/factory_builder.go]: Thanks for adding the comment. It looks good now. As per test-run and removing the label, @mengqiy took care of it.

File: pkg/printers/printers.go [hunk 3]
```diff
@@ -25,92 +25,102 @@ import (
...
-			return nil, false, err
+			return nil, err
 		}
+
 	case "wide":
 		fallthrough
 	case "":
-		return nil, false, nil
+
+		printer = NewHumanReadablePrinter(encoder, decoders[0], options)
```
Comment (mengqiy) [inline pkg/printers/printers.go]: Do we need to move this line up for `wide` only?
I suspect it changes the behavior when not specify `--output` flag.

Comment (waseem) [inline pkg/printers/printers.go]: @mengqiy I don't follow how that is possible.

There are four general cases for `--output` flag.

1. It's one of the generic formats these are `json`, `yaml`, `name`, `template/go-template`, `templatefile/go-templatefile`, `jsonpath`, `jsonpath-file`, `custom-columns`, or `custom-columns-file`.
2. A non-generic format `wide` or `""`.
3. An unknown format.

In the context of just this function, `wide` and `""` are treated the same because of the blank `fallthrough` in the case of `wide`. We could also have used just `case "wide", "":` to get the same result.

Now in case `--output` is actually set to `""`, it is actually set to `template` [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/printing.go#L111,L118) and then it is used to retrieve the printer.

Should we get another pair of eyes for this?

**Comment by fabianofranz** [comment]
@waseem mind squashing your commits?

**Comment by waseem** [comment]
@fabianofranz I squashed the commits and rebased the branch.

**Comment by fabianofranz** [comment]
@waseem thanks!
/approve
/lgtm

**Comment by fabianofranz** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-unit test this

**Comment by fabianofranz** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
